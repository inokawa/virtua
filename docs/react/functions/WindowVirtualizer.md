[**API**](../../API.md) • **Docs**

***

# Function: WindowVirtualizer()

> **WindowVirtualizer**(`props`): `ReactNode`

[Virtualizer](Virtualizer.md) controlled by the window scrolling. See [WindowVirtualizerProps](../interfaces/WindowVirtualizerProps.md) and [WindowVirtualizer](WindowVirtualizer.md).

## Parameters

• **props**: [`WindowVirtualizerProps`](../interfaces/WindowVirtualizerProps.md) & `RefAttributes`\<[`WindowVirtualizerHandle`](../interfaces/WindowVirtualizerHandle.md)\>

## Returns

`ReactNode`

## Defined in

[src/react/WindowVirtualizer.tsx:117](https://github.com/inokawa/virtua/blob/21b5b4e90d3d1661add9e7a1149339bf1d18958e/src/react/WindowVirtualizer.tsx#L117)
