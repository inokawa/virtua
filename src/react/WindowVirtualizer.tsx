import {
  ReactElement,
  ReactNode,
  useEffect,
  forwardRef,
  useImperativeHandle,
  useRef,
} from "react";
import {
  ACTION_ITEMS_LENGTH_CHANGE,
  UPDATE_VIRTUAL_STATE,
  getOverscanedRange,
  createVirtualStore,
  SCROLL_IDLE,
  UPDATE_SCROLL_END_EVENT,
} from "../core/store";
import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect";
import { createWindowScroller } from "../core/scroller";
import { getKey, refKey } from "./utils";
import { useStatic } from "./useStatic";
import { useLatestRef } from "./useLatestRef";
import { createWindowResizer } from "../core/resizer";
import { CacheSnapshot } from "../core/types";
import { CustomContainerComponent, CustomItemComponent } from "./types";
import { ListItem } from "./ListItem";
import { flushSync } from "react-dom";
import { useRerender } from "./useRerender";
import { useChildren } from "./useChildren";

/**
 * Methods of {@link WindowVirtualizer}.
 */
export interface WindowVirtualizerHandle {
  /**
   * Get current {@link CacheSnapshot}.
   */
  readonly cache: CacheSnapshot;
}

/**
 * Props of {@link WindowVirtualizer}.
 */
export interface WindowVirtualizerProps {
  /**
   * Elements rendered by this component.
   *
   * You can also pass a function and set {@link WindowVirtualizerProps.count} to create elements lazily.
   */
  children: ReactNode | ((index: number) => ReactElement);
  /**
   * If you set a function to {@link WindowVirtualizerProps.children}, you have to set total number of items to this prop.
   */
  count?: number;
  /**
   * Number of items to render above/below the visible bounds of the list. Lower value will give better performance but you can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan?: number;
  /**
   * Item size hint for unmeasured items. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize?: number;
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift?: boolean;
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal?: boolean;
  /**
   * You can restore cache by passing a {@link CacheSnapshot} on mount. This is useful when you want to restore scroll position after navigation. The snapshot can be obtained from {@link WindowVirtualizerHandle.cache}.
   *
   * **The length of items should be the same as when you take the snapshot, otherwise restoration may not work as expected.**
   */
  cache?: CacheSnapshot;
  /**
   * A prop for SSR. If set, the specified amount of items will be mounted in the initial rendering regardless of the container size until hydrated.
   */
  ssrCount?: number;
  /**
   * Component or element type for container element.
   * @defaultValue "div"
   */
  as?: keyof JSX.IntrinsicElements | CustomContainerComponent;
  /**
   * Component or element type for item element. This component will get {@link CustomItemComponentProps} as props.
   * @defaultValue "div"
   */
  item?: keyof JSX.IntrinsicElements | CustomItemComponent;
  /**
   * Callback invoked when scrolling stops.
   */
  onScrollEnd?: () => void;
  /**
   * Callback invoked when visible items range changes.
   */
  onRangeChange?: (
    /**
     * The start index of viewable items.
     */
    startIndex: number,
    /**
     * The end index of viewable items.
     */
    endIndex: number
  ) => void;
}

/**
 * {@link Virtualizer} controlled by the window scrolling. See {@link WindowVirtualizerProps} and {@link WindowVirtualizer}.
 */
export const WindowVirtualizer = forwardRef<
  WindowVirtualizerHandle,
  WindowVirtualizerProps
>(
  (
    {
      children,
      count: renderCountProp,
      overscan = 4,
      itemSize,
      shift,
      horizontal: horizontalProp,
      cache,
      ssrCount,
      as: Element = "div",
      item: ItemElement = "div",
      onScrollEnd: onScrollEndProp,
      onRangeChange: onRangeChangeProp,
    },
    ref
  ): ReactElement => {
    Element = Element as "div";

    const [getElement, count] = useChildren(children, renderCountProp);

    const containerRef = useRef<HTMLDivElement>(null);

    const onScrollEnd = useLatestRef(onScrollEndProp);

    const isSSR = useRef(!!ssrCount);

    const [store, resizer, scroller, isHorizontal] = useStatic(() => {
      const _isHorizontal = !!horizontalProp;
      const _store = createVirtualStore(
        count,
        itemSize,
        ssrCount,
        cache,
        !itemSize
      );

      return [
        _store,
        createWindowResizer(_store, _isHorizontal),
        createWindowScroller(_store, _isHorizontal),
        _isHorizontal,
      ];
    });
    // The elements length and cached items length are different just after element is added/removed.
    if (count !== store._getItemsLength()) {
      store._update(ACTION_ITEMS_LENGTH_CHANGE, [count, shift]);
    }

    const rerender = useRerender(store);

    const [startIndex, endIndex] = store._getRange();
    const scrollDirection = store._getScrollDirection();
    const jumpCount = store._getJumpCount();
    const totalSize = store._getTotalSize();

    const items: ReactElement[] = [];

    useIsomorphicLayoutEffect(() => {
      isSSR[refKey] = false;

      // store must be subscribed first because others may dispatch update on init depending on implementation
      const unsubscribeStore = store._subscribe(
        UPDATE_VIRTUAL_STATE,
        (sync) => {
          if (sync) {
            flushSync(rerender);
          } else {
            rerender();
          }
        }
      );
      const unsubscribeOnScrollEnd = store._subscribe(
        UPDATE_SCROLL_END_EVENT,
        () => {
          onScrollEnd[refKey] && onScrollEnd[refKey]();
        }
      );

      const el = containerRef[refKey]!;
      resizer._observeRoot(el);
      scroller._observe(el);
      return () => {
        unsubscribeStore();
        unsubscribeOnScrollEnd();
        resizer._dispose();
        scroller._dispose();
      };
    }, []);

    useIsomorphicLayoutEffect(() => {
      scroller._fixScrollJump();
    }, [jumpCount]);

    useEffect(() => {
      if (!onRangeChangeProp) return;

      onRangeChangeProp(startIndex, endIndex);
    }, [startIndex, endIndex]);

    useImperativeHandle(
      ref,
      () => {
        return {
          get cache() {
            return store._getCacheSnapshot();
          },
        };
      },
      []
    );

    for (
      let [i, j] = getOverscanedRange(
        startIndex,
        endIndex,
        overscan,
        scrollDirection,
        count
      );
      i <= j;
      i++
    ) {
      const e = getElement(i);
      items.push(
        <ListItem
          key={getKey(e, i)}
          _resizer={resizer._observeItem}
          _index={i}
          _offset={store._getItemOffset(i)}
          _hide={store._isUnmeasuredItem(i)}
          _as={ItemElement as "div"}
          _children={e}
          _isHorizontal={isHorizontal}
          _isSSR={isSSR[refKey]}
        />
      );
    }

    return (
      <Element
        ref={containerRef}
        style={{
          // contain: "content",
          flex: "none", // flex style can break layout
          position: "relative",
          visibility: "hidden", // TODO replace with other optimization methods
          width: isHorizontal ? totalSize : "100%",
          height: isHorizontal ? "100%" : totalSize,
          pointerEvents: scrollDirection !== SCROLL_IDLE ? "none" : "auto",
        }}
      >
        {items}
      </Element>
    );
  }
);
