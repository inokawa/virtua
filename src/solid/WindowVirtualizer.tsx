/**
 * @jsxImportSource solid-js
 */
import {
  onMount,
  onCleanup,
  createEffect,
  createSignal,
  createMemo,
  JSX,
  on,
  createComputed,
} from "solid-js";
import {
  SCROLL_IDLE,
  UPDATE_SCROLL_END_EVENT,
  UPDATE_VIRTUAL_STATE,
  getOverscanedRange,
  createVirtualStore,
  ACTION_ITEMS_LENGTH_CHANGE,
} from "../core/store";
import { createWindowResizer } from "../core/resizer";
import { createWindowScroller } from "../core/scroller";
import { ListItem } from "./ListItem";
import { RangedFor } from "./RangedFor";
import { isSameRange } from "./utils";
import { ItemsRange } from "../core/types";

// /**
//  * Methods of {@link WindowVirtualizer}.
//  */
// export interface WindowVirtualizerHandle {}

/**
 * Props of {@link WindowVirtualizer}.
 */
export interface WindowVirtualizerProps<T> {
  // /**
  //  * Get reference to {@link WindowVirtualizerHandle}.
  //  */
  // ref?: (handle?: WindowVirtualizerHandle) => void;
  /**
   * The data items rendered by this component.
   */
  data: T[];
  /**
   * The elements renderer function.
   */
  children: (data: T, index: number) => JSX.Element;
  /**
   * Number of items to render above/below the visible bounds of the list. Lower value will give better performance but you can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan?: number;
  /**
   * Item size hint for unmeasured items. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize?: number;
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift?: boolean;
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal?: boolean;
  /**
   * Callback invoked when scrolling stops.
   */
  onScrollEnd?: () => void;
  /**
   * Callback invoked when visible items range changes.
   */
  onRangeChange?: (
    /**
     * The start index of viewable items.
     */
    startIndex: number,
    /**
     * The end index of viewable items.
     */
    endIndex: number
  ) => void;
}

/**
 * {@link Virtualizer} controlled by the window scrolling. See {@link WindowVirtualizerProps} and {@link WindowVirtualizer}.
 */
export const WindowVirtualizer = <T,>(
  props: WindowVirtualizerProps<T>
): JSX.Element => {
  let containerRef: HTMLDivElement | undefined;

  const {
    // ref: _ref,
    data: _data,
    children: _children,
    overscan: _overscan,
    itemSize,
    shift: _shift,
    horizontal = false,
    onScrollEnd: _onScrollEnd,
    onRangeChange: _onRangeChange,
  } = props;

  const store = createVirtualStore(
    props.data.length,
    itemSize ?? 40,
    undefined,
    undefined,
    !itemSize
  );
  const resizer = createWindowResizer(store, horizontal);
  const scroller = createWindowScroller(store, horizontal);

  const [rerender, setRerender] = createSignal(store._getStateVersion());

  const unsubscribeStore = store._subscribe(UPDATE_VIRTUAL_STATE, () => {
    setRerender(store._getStateVersion());
  });

  const unsubscribeOnScrollEnd = store._subscribe(
    UPDATE_SCROLL_END_EVENT,
    () => {
      props.onScrollEnd?.();
    }
  );

  const range = createMemo<ItemsRange>((prev) => {
    rerender();
    const next = store._getRange();
    if (prev && isSameRange(prev, next)) {
      return prev;
    }
    return next;
  });
  const scrollDirection = createMemo(
    () => rerender() && store._getScrollDirection()
  );
  const totalSize = createMemo(() => rerender() && store._getTotalSize());

  const jumpCount = createMemo(() => rerender() && store._getJumpCount());

  const overscanedRange = createMemo<ItemsRange>((prev) => {
    const overscan = props.overscan ?? 4;
    const [startIndex, endIndex] = range();
    const next = getOverscanedRange(
      startIndex,
      endIndex,
      overscan,
      scrollDirection(),
      props.data.length
    );
    if (prev && isSameRange(prev, next)) {
      return prev;
    }
    return next;
  });

  onMount(() => {
    resizer._observeRoot(containerRef!);
    scroller._observe(containerRef!);

    onCleanup(() => {
      unsubscribeStore();
      unsubscribeOnScrollEnd();
      resizer._dispose();
      scroller._dispose();
    });
  });

  createComputed(
    on(
      () => props.data.length,
      (len) => {
        if (len !== store._getItemsLength()) {
          store._update(ACTION_ITEMS_LENGTH_CHANGE, [len, props.shift]);
        }
      }
    )
  );

  createEffect(
    on(jumpCount, () => {
      scroller._fixScrollJump();
    })
  );

  createEffect(() => {
    const next = range();
    props.onRangeChange && props.onRangeChange(next[0], next[1]);
  });

  return (
    <div
      ref={containerRef}
      style={{
        // contain: "content",
        "overflow-anchor": "none", // opt out browser's scroll anchoring because it will conflict to scroll anchoring of virtualizer
        flex: "none", // flex style can break layout
        position: "relative",
        visibility: "hidden", // TODO replace with other optimization methods
        width: horizontal ? totalSize() + "px" : "100%",
        height: horizontal ? "100%" : totalSize() + "px",
        "pointer-events":
          scrollDirection() !== SCROLL_IDLE ? "none" : undefined,
      }}
    >
      <RangedFor
        _each={props.data}
        _range={overscanedRange()}
        _render={(data, index) => {
          const offset = createMemo(() => {
            rerender();
            return store._getItemOffset(index);
          });
          const hide = createMemo(() => {
            rerender();
            return store._isUnmeasuredItem(index);
          });

          return (
            <ListItem
              _index={index}
              _resizer={resizer._observeItem}
              _offset={offset()}
              _hide={hide()}
              _children={props.children(data(), index)}
              _isHorizontal={horizontal}
            />
          );
        }}
      />
    </div>
  );
};
