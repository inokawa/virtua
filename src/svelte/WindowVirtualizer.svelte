<script lang="ts" generics="T">
  import { onMount, onDestroy } from "svelte";
  import {
    type StateVersion,
    CHANGE_ITEM_LENGTH,
    createWindowVirtualizer,
    FIX_SCROLL_JUMP,
    GET_ITEM_OFFSET,
    GET_JUMP_COUNT,
    GET_RANGE,
    GET_IS_SCROLLING,
    GET_TOTAL_SIZE,
    IS_ITEM_HIDDEN,
    OBSERVE_ITEM_RESIZE,
    ON_MOUNT,
    ON_UN_MOUNT,
    GET_ITEMS_LENGTH,
    FIND_START_INDEX,
    FIND_END_INDEX,
  } from "./core";
  import { defaultGetKey, iterRange, styleToString } from "./utils";
  import ListItem from "./ListItem.svelte";
  import type {
    WindowVirtualizerHandle,
    WindowVirtualizerProps,
  } from "./WindowVirtualizer.type";

  interface Props extends WindowVirtualizerProps<T> {}

  let {
    data,
    getKey = defaultGetKey,
    overscan,
    itemSize,
    shift = false,
    horizontal = false,
    children,
    onscroll,
    onscrollend,
  }: Props = $props();

  const virtualizer = createWindowVirtualizer(
    data.length,
    itemSize,
    overscan,
    horizontal,
    (v) => {
      rerender = v;
    },
    (offset) => {
      onscroll && onscroll(offset);
    },
    () => {
      onscrollend && onscrollend();
    }
  );

  let containerRef: HTMLDivElement | undefined = $state();

  let rerender: StateVersion = $state([]);

  let range = $derived(rerender && virtualizer[GET_RANGE]());
  let isScrolling = $derived(rerender && virtualizer[GET_IS_SCROLLING]());
  let totalSize = $derived(rerender && virtualizer[GET_TOTAL_SIZE]());
  let jumpCount = $derived(rerender && virtualizer[GET_JUMP_COUNT]());

  onMount(() => {
    virtualizer[ON_MOUNT](containerRef!);
  });
  onDestroy(() => {
    virtualizer[ON_UN_MOUNT]();
  });

  $effect.pre(() => {
    if (data.length !== virtualizer[GET_ITEMS_LENGTH]()) {
      virtualizer[CHANGE_ITEM_LENGTH](data.length, shift);
    }
  });

  let prevJumpCount: number | undefined;
  $effect(() => {
    if (prevJumpCount === jumpCount) return;
    prevJumpCount = jumpCount;
    virtualizer[FIX_SCROLL_JUMP]();
  });

  export const findStartIndex = virtualizer[
    FIND_START_INDEX
  ] satisfies WindowVirtualizerHandle["findStartIndex"] as WindowVirtualizerHandle["findStartIndex"];
  export const findEndIndex = virtualizer[
    FIND_END_INDEX
  ] satisfies WindowVirtualizerHandle["findEndIndex"] as WindowVirtualizerHandle["findEndIndex"];

  let containerStyle = $derived(
    styleToString({
      // contain: "content",
      "overflow-anchor": "none", // opt out browser's scroll anchoring because it will conflict to scroll anchoring of virtualizer
      flex: "none", // flex style can break layout
      position: "relative",
      visibility: "hidden", // TODO replace with other optimization methods
      width: horizontal ? totalSize + "px" : "100%",
      height: horizontal ? "100%" : totalSize + "px",
      "pointer-events": isScrolling ? "none" : undefined,
    })
  );
</script>

<!-- 
  @component
  {@link Virtualizer} controlled by the window scrolling. See {@link WindowVirtualizerProps} and {@link WindowVirtualizerHandle}.
-->
<div bind:this={containerRef} style={containerStyle}>
  {#each iterRange(range) as index (getKey(data[index]!, index))}
    {@const item = data[index]!}
    <ListItem
      {children}
      {item}
      {index}
      as={"div"}
      offset={rerender && virtualizer[GET_ITEM_OFFSET](index)}
      hide={rerender && virtualizer[IS_ITEM_HIDDEN](index)}
      {horizontal}
      resizer={virtualizer[OBSERVE_ITEM_RESIZE]}
    />
  {/each}
</div>
