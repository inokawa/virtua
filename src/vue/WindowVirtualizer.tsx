/** @jsxImportSource vue */
import {
  ref,
  onMounted,
  defineComponent,
  onUnmounted,
  VNode,
  watch,
  ComponentOptionsMixin,
  SlotsType,
  ComponentOptionsWithObjectProps,
  ComponentObjectPropsOptions,
  PropType,
  NativeElements,
  computed,
} from "vue";
import {
  UPDATE_SCROLL_END_EVENT,
  UPDATE_VIRTUAL_STATE,
  createVirtualStore,
  ACTION_ITEMS_LENGTH_CHANGE,
  UPDATE_SCROLL_EVENT,
} from "../core/store";
import { createWindowResizer } from "../core/resizer";
import { createWindowScroller } from "../core/scroller";
import { ListItem } from "./ListItem";
import { getKey, isSameRange } from "./utils";
import { ItemsRange, ScrollToIndexOpts } from "../core/types";

export interface WindowVirtualizerHandle {
  /**
   * Find the start index of visible range of items.
   */
  findStartIndex: () => number;
  /**
   * Find the end index of visible range of items.
   */
  findEndIndex: () => number;
  /**
   * Scroll to the item specified by index.
   * @param index index of item
   * @param opts options
   */
  scrollToIndex(index: number, opts?: ScrollToIndexOpts): void;
}

const props = {
  /**
   * The data items rendered by this component.
   */
  data: { type: Array, required: true },
  /**
   * Number of items to render above/below the visible bounds of the list. You can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan: Number,
  /**
   * Item size hint for unmeasured items. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize: Number,
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift: Boolean,
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal: Boolean,
  /**
   * Component or element type for container element.
   * @defaultValue "div"
   */
  as: { type: String as PropType<keyof NativeElements>, default: "div" },
  /**
   * Component or element type for item element.
   * @defaultValue "div"
   */
  item: { type: String as PropType<keyof NativeElements>, default: "div" },
} satisfies ComponentObjectPropsOptions;

export const WindowVirtualizer = /*#__PURE__*/ defineComponent({
  props,
  emits: ["scroll", "scrollEnd"],
  setup(props, { emit, slots, expose }) {
    const isHorizontal = props.horizontal;
    const containerRef = ref<HTMLDivElement>();
    const store = createVirtualStore(
      props.data.length,
      props.itemSize,
      props.overscan,
      undefined,
      undefined,
      !props.itemSize
    );
    const resizer = createWindowResizer(store, isHorizontal);
    const scroller = createWindowScroller(store, isHorizontal);

    const rerender = ref(store._getStateVersion());
    const unsubscribeStore = store._subscribe(UPDATE_VIRTUAL_STATE, () => {
      rerender.value = store._getStateVersion();
    });

    const unsubscribeOnScroll = store._subscribe(UPDATE_SCROLL_EVENT, () => {
      emit("scroll", store._getScrollOffset());
    });
    const unsubscribeOnScrollEnd = store._subscribe(
      UPDATE_SCROLL_END_EVENT,
      () => {
        emit("scrollEnd");
      }
    );

    const range = computed<ItemsRange>((prev) => {
      rerender.value;
      const next = store._getRange();
      if (prev && isSameRange(prev, next)) {
        return prev;
      }
      return next;
    });
    const isScrolling = computed(() => rerender.value && store._isScrolling());
    const totalSize = computed(() => rerender.value && store._getTotalSize());
    const jumpCount = computed(() => rerender.value && store._getJumpCount());

    onMounted(() => {
      const el = containerRef.value;
      if (!el) return;
      resizer._observeRoot(el);
      scroller._observe(el);
    });
    onUnmounted(() => {
      unsubscribeStore();
      unsubscribeOnScroll();
      unsubscribeOnScrollEnd();
      resizer._dispose();
      scroller._dispose();
    });

    watch(
      () => props.data.length,
      (count) => {
        store._update(ACTION_ITEMS_LENGTH_CHANGE, [count, props.shift]);
      }
    );

    watch(
      [jumpCount],
      () => {
        scroller._fixScrollJump();
      },
      { flush: "post" }
    );

    expose({
      findStartIndex: store._findStartIndex,
      findEndIndex: store._findEndIndex,
      scrollToIndex: scroller._scrollToIndex,
    } satisfies WindowVirtualizerHandle);

    return () => {
      const Element = props.as;
      const ItemElement = props.item;

      const [startIndex, endIndex] = range.value;
      const total = totalSize.value;

      const items: VNode[] = [];
      for (let i = startIndex, j = endIndex; i <= j; i++) {
        const e = slots.default({ item: props.data![i]!, index: i })[0]!;
        items.push(
          <ListItem
            key={getKey(e, i)}
            _rerender={rerender}
            _store={store}
            _resizer={resizer._observeItem}
            _index={i}
            _children={e}
            _isHorizontal={isHorizontal}
            _as={ItemElement}
          />
        );
      }

      return (
        <Element
          ref={containerRef}
          style={{
            // contain: "content",
            overflowAnchor: "none", // opt out browser's scroll anchoring because it will conflict to scroll anchoring of virtualizer
            flex: "none", // flex style can break layout
            position: "relative",
            visibility: "hidden", // TODO replace with other optimization methods
            width: isHorizontal ? total + "px" : "100%",
            height: isHorizontal ? "100%" : total + "px",
            pointerEvents: isScrolling.value ? "none" : undefined,
          }}
        >
          {items}
        </Element>
      );
    };
  },
} as ComponentOptionsWithObjectProps<
  typeof props,
  void,
  {},
  {},
  {},
  ComponentOptionsMixin,
  ComponentOptionsMixin,
  {
    /**
     * Callback invoked whenever scroll offset changes.
     * @param offset Current scrollTop, or scrollLeft if horizontal: true.
     */
    scroll: (offset: number) => void;
    /**
     * Callback invoked when scrolling stops.
     */
    scrollEnd: () => void;
  },
  string,
  {},
  string,
  SlotsType<{ default: (arg: { item: any; index: number }) => VNode[] }>
>);
