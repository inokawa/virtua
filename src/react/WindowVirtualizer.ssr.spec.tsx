/**
 * @vitest-environment node
 */
import { it, describe, expect } from "vitest";
import { renderToString, renderToStaticMarkup } from "react-dom/server";
import { WindowVirtualizer } from "./WindowVirtualizer.js";
import { JSDOM } from "jsdom";

const LIST_ID = "list-id";

describe("SSR", () => {
  it("should render items with renderToString and vertical", () => {
    const COUNT = 10;
    const OVERSCAN = 4;
    const html = renderToString(
      <div id={LIST_ID}>
        <WindowVirtualizer ssrCount={COUNT} overscan={OVERSCAN}>
          {Array.from({ length: 1000 }).map((_, i) => (
            <div key={i}>{i}</div>
          ))}
        </WindowVirtualizer>
      </div>
    );
    expect(html).toMatchSnapshot();

    expect(
      new JSDOM(html).window.document.getElementById(LIST_ID)!
        .firstElementChild!.childElementCount
    ).toEqual(COUNT + OVERSCAN);
  });

  it("should render items with renderToStaticMarkup and vertical", () => {
    const COUNT = 10;
    const OVERSCAN = 4;
    const html = renderToStaticMarkup(
      <div id={LIST_ID}>
        <WindowVirtualizer ssrCount={COUNT} overscan={OVERSCAN}>
          {Array.from({ length: 1000 }).map((_, i) => (
            <div key={i}>{i}</div>
          ))}
        </WindowVirtualizer>
      </div>
    );
    expect(html).toMatchSnapshot();

    expect(
      new JSDOM(html).window.document.getElementById(LIST_ID)!
        .firstElementChild!.childElementCount
    ).toEqual(COUNT + OVERSCAN);
  });

  it("should render items with renderToString and horizontal", () => {
    const COUNT = 10;
    const OVERSCAN = 4;
    const html = renderToString(
      <div id={LIST_ID}>
        <WindowVirtualizer
          ssrCount={COUNT}
          overscan={OVERSCAN}
          horizontal
        >
          {Array.from({ length: 1000 }).map((_, i) => (
            <div key={i}>{i}</div>
          ))}
        </WindowVirtualizer>
      </div>
    );
    expect(html).toMatchSnapshot();

    expect(
      new JSDOM(html).window.document.getElementById(LIST_ID)!
        .firstElementChild!.childElementCount
    ).toEqual(COUNT + OVERSCAN);
  });

  it("should render items with renderToStaticMarkup and horizontal", () => {
    const COUNT = 10;
    const OVERSCAN = 4;
    const html = renderToStaticMarkup(
      <div id={LIST_ID}>
        <WindowVirtualizer
          ssrCount={COUNT}
          overscan={OVERSCAN}
          horizontal
        >
          {Array.from({ length: 1000 }).map((_, i) => (
            <div key={i}>{i}</div>
          ))}
        </WindowVirtualizer>
      </div>
    );
    expect(html).toMatchSnapshot();

    expect(
      new JSDOM(html).window.document.getElementById(LIST_ID)!
        .firstElementChild!.childElementCount
    ).toEqual(COUNT + OVERSCAN);
  });
});
