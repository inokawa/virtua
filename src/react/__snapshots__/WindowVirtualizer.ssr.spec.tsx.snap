// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SSR > should render items with renderToStaticMarkup and horizontal 1`] = `"<div id="list-id"><div style="flex:none;position:relative;visibility:hidden;width:40000px;height:100%"><div style="height:100%;top:0;left:0;visibility:visible;display:flex"><div>0</div></div><div style="height:100%;top:0;left:40px;visibility:visible;display:flex"><div>1</div></div><div style="height:100%;top:0;left:80px;visibility:visible;display:flex"><div>2</div></div><div style="height:100%;top:0;left:120px;visibility:visible;display:flex"><div>3</div></div><div style="height:100%;top:0;left:160px;visibility:visible;display:flex"><div>4</div></div><div style="height:100%;top:0;left:200px;visibility:visible;display:flex"><div>5</div></div><div style="height:100%;top:0;left:240px;visibility:visible;display:flex"><div>6</div></div><div style="height:100%;top:0;left:280px;visibility:visible;display:flex"><div>7</div></div><div style="height:100%;top:0;left:320px;visibility:visible;display:flex"><div>8</div></div><div style="height:100%;top:0;left:360px;visibility:visible;display:flex"><div>9</div></div><div style="height:100%;top:0;left:400px;visibility:visible;display:flex"><div>10</div></div><div style="height:100%;top:0;left:440px;visibility:visible;display:flex"><div>11</div></div><div style="height:100%;top:0;left:480px;visibility:visible;display:flex"><div>12</div></div><div style="height:100%;top:0;left:520px;visibility:visible;display:flex"><div>13</div></div></div></div>"`;

exports[`SSR > should render items with renderToStaticMarkup and vertical 1`] = `"<div id="list-id"><div style="flex:none;position:relative;visibility:hidden;width:100%;height:40000px"><div style="width:100%;left:0;top:0;visibility:visible"><div>0</div></div><div style="width:100%;left:0;top:40px;visibility:visible"><div>1</div></div><div style="width:100%;left:0;top:80px;visibility:visible"><div>2</div></div><div style="width:100%;left:0;top:120px;visibility:visible"><div>3</div></div><div style="width:100%;left:0;top:160px;visibility:visible"><div>4</div></div><div style="width:100%;left:0;top:200px;visibility:visible"><div>5</div></div><div style="width:100%;left:0;top:240px;visibility:visible"><div>6</div></div><div style="width:100%;left:0;top:280px;visibility:visible"><div>7</div></div><div style="width:100%;left:0;top:320px;visibility:visible"><div>8</div></div><div style="width:100%;left:0;top:360px;visibility:visible"><div>9</div></div><div style="width:100%;left:0;top:400px;visibility:visible"><div>10</div></div><div style="width:100%;left:0;top:440px;visibility:visible"><div>11</div></div><div style="width:100%;left:0;top:480px;visibility:visible"><div>12</div></div><div style="width:100%;left:0;top:520px;visibility:visible"><div>13</div></div></div></div>"`;

exports[`SSR > should render items with renderToString and horizontal 1`] = `"<div id="list-id"><div style="flex:none;position:relative;visibility:hidden;width:40000px;height:100%"><div style="height:100%;top:0;left:0;visibility:visible;display:flex"><div>0</div></div><div style="height:100%;top:0;left:40px;visibility:visible;display:flex"><div>1</div></div><div style="height:100%;top:0;left:80px;visibility:visible;display:flex"><div>2</div></div><div style="height:100%;top:0;left:120px;visibility:visible;display:flex"><div>3</div></div><div style="height:100%;top:0;left:160px;visibility:visible;display:flex"><div>4</div></div><div style="height:100%;top:0;left:200px;visibility:visible;display:flex"><div>5</div></div><div style="height:100%;top:0;left:240px;visibility:visible;display:flex"><div>6</div></div><div style="height:100%;top:0;left:280px;visibility:visible;display:flex"><div>7</div></div><div style="height:100%;top:0;left:320px;visibility:visible;display:flex"><div>8</div></div><div style="height:100%;top:0;left:360px;visibility:visible;display:flex"><div>9</div></div><div style="height:100%;top:0;left:400px;visibility:visible;display:flex"><div>10</div></div><div style="height:100%;top:0;left:440px;visibility:visible;display:flex"><div>11</div></div><div style="height:100%;top:0;left:480px;visibility:visible;display:flex"><div>12</div></div><div style="height:100%;top:0;left:520px;visibility:visible;display:flex"><div>13</div></div></div></div>"`;

exports[`SSR > should render items with renderToString and vertical 1`] = `"<div id="list-id"><div style="flex:none;position:relative;visibility:hidden;width:100%;height:40000px"><div style="width:100%;left:0;top:0;visibility:visible"><div>0</div></div><div style="width:100%;left:0;top:40px;visibility:visible"><div>1</div></div><div style="width:100%;left:0;top:80px;visibility:visible"><div>2</div></div><div style="width:100%;left:0;top:120px;visibility:visible"><div>3</div></div><div style="width:100%;left:0;top:160px;visibility:visible"><div>4</div></div><div style="width:100%;left:0;top:200px;visibility:visible"><div>5</div></div><div style="width:100%;left:0;top:240px;visibility:visible"><div>6</div></div><div style="width:100%;left:0;top:280px;visibility:visible"><div>7</div></div><div style="width:100%;left:0;top:320px;visibility:visible"><div>8</div></div><div style="width:100%;left:0;top:360px;visibility:visible"><div>9</div></div><div style="width:100%;left:0;top:400px;visibility:visible"><div>10</div></div><div style="width:100%;left:0;top:440px;visibility:visible"><div>11</div></div><div style="width:100%;left:0;top:480px;visibility:visible"><div>12</div></div><div style="width:100%;left:0;top:520px;visibility:visible"><div>13</div></div></div></div>"`;
